working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,884 @@
+
+*In particular, it is allowed to create a new instance where one or more
+extension types occur as type arguments.*
+
+When `k` is zero, `Ext<S1, .. Sk>` simply stands for `Ext`, a non-generic
+extension. When `k` is greater than zero, a raw occurrence `Ext` is treated
+like a raw type: Instantiation to bound is used to obtain the omitted type
+arguments.
+
+We say that the static type of said variable, parameter, etc. _is the
+extension type_ `Ext<S1, .. Sk>`, and that its static type _is an extension
+type_.
+
+If `e` is an expression whose static type is the extension type
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>></code>,
+then a member access like `e.m(args)` is treated as
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>>(v).m(args)</code>
+where `v` is a fresh variable whose declared type
+is the on-type corresponding to `Ext<S1, .. Sm>`, and similarly for
+instance getters and operators. This rule also applies when a member access
+implicitly has the receiver `this`, and the static type of `this` is an
+extension type (*which can only occur in an extension type member
+declaration*).
+
+*That is, when the type of an expression is an extension type, all method
+invocations on that expression will invoke an extension method declared by
+that extension, and similarly for other member accesses. In particular, we
+cannot invoke an instance member when the receiver type is an extension
+type (unless the the extension type enables them explicitly, cf. the
+show/hide part specified in a later section).*
+
+Let `E` be an extension declaration named `Ext` with type parameters
+<code>X<sub>1</sub> extends B<sub>1</sub>, .. X<sub>k</sub> extends B<sub>k</sub></code>
+and on-type clause `on T`. Then we say that the _declared on-type_ of `Ext`
+is `T`, and the _instantiated on-type_ of 
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>></code> 
+is
+<code>[S<sub>1</sub>/X<sub>1</sub>, .. S<sub>k</sub>/X<sub>k</sub>]T</code>.
+We will omit 'declared' and 'instantiated' from the phrase when it is clear
+from the context whether we are talking about the extension itself or a
+particular instantiation of a generic extension. For non-generic
+extensions, the on-type is the same in either case.
+
+An extension type `E` of the form 
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>></code>
+is a subtype of `Object?`, and a proper supertype of the
+instantiated on-type of `E`. In the case where the instantiated on-type of
+`E` is not a top type, `E` is also a proper subtype of `Object?`.
+
+*That is, the underlying on-type can only be recovered by an explicit cast
+(except when the on-type is a top type). So an expression whose type is an
+extension type is in a sense "in prison", and we can only obtain a
+different type for it by forgetting everything (going to a top type), or by
+means of an explicit cast, typically a downcast to the on-type.*
+
+*There is one exception where an extension type has a supertype which is
```

Comment:
I don't think this makes a lot of sense to me.  

First, nit, but - I think you need to be very explicit about instantiated here - `...is an extension type with instantiated on-type...` .  

Second... this feels like a super weird exception to me.  Why carve out this one out?  I don't understand the argument about "disciplines".   I can extend other extensions for purposes having nothing to do with the other extension, why should I be able to freely assign between them?

Third, actually, I don't understand this at all.  This isn't an exception at all, right?  This is just another instance of the principle you listed above (which I objected to) "It illustrates the fact that an extension type E with on-type T introduces a type E which is a supertype of T." 

So I don't think this aside is saying anything except what you've already said above.

Reply:
Here's the scenario I'm thinking about:

```dart
extension type E2 on int show num {}
extension type E1 on E2 {
  E1 nextFloor() => this.floor() + 1;
  String get hello => 'Hello!';
}

void main() {
  E2 e2 = 42; // OK, upcast.
  E1 e1 = e2; // OK, upcast.

  // Imposed discipline: Only `num` members allowed.
  e2.isEven; // Error.
  e2.floor(); // OK.

  // Even stricter discipline, even with the additional members.
  e1.floor(); // Error.
  e1.nextFloor(); // OK, we can use `floor`, but only as allowed by E1.
}
```

Of course, a receiver of type `E1` allows for a larger set of members, but each of them is implemented using members that are available on `E2`, and hence they provide an even more disciplined use of the members of `int` than `E1` does.

We can easily break this abstraction by a cast to `int`, but the abstraction is available, and it is possible to use it as a developer who doesn't actually _want_ to violate the encoded discipline.

Reply:
I'm completely lost.  Nothing you wrote there is surprising to me given the text you wrote previously up to this point.  But you start this paragraph with "There is one exception...".  What exception are you introducing?  What part of the code above would change if you deleted this paragraph?

Reply:
I also think that this is confusing more than it's explaining. There is some underlying model behind this, one which isn't explained (or not explained in terms that makes it clear that it's the base model that the "exception" should be taken against). Is it the "in prison"/"only way to escape is to go to a top type" thing? Then just change that to "up-cast to a super-type" instead, and then this is an example, not an exception.

An extension type is a subtype of `Object?` and a supertype of its `on`-type.
That explains everything. If the `on` type is an extension type, then, well, everything still follows from that rule. No exception needed.

It also seems like there is a particular use-case behind this explanation, but use-cases are examples, not explanations. There will be people for whom that use-case never become relevant, and everything should make sense to them anyway without that example.

Reply:
Deleted the paragraph.